### PR TITLE
Tracks configure map

### PR DESCRIPTION
**Pull request by DmitrySvetlichny**

### CODE REVIEW

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 1]
```diff
@@ -0,0 +1,592 @@
...
+    private let previouslyVisibleTracksKey: String = "PreviouslyVisibleGpxFilePaths"
+    private var searchController: UISearchController?
+    private var segmentedControl: UISegmentedControl?
+    private var allGpxList: [OAGPX] = []
+    private var visibleGpxList: [OAGPX] = []
+    private var recentlyVisibleGpxList: [OAGPX] = []
+    private var filteredGpxList: [OAGPX] = []
+    private var selectedGpxTracks: [OAGPX] = []
+    private var previousSelectedSegmentIndex: Int = 0
+    private var isShowingVisibleTracks: Bool = true
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:24]: private var isShowingVisibleTracks = true (and other)

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 2]
```diff
@@ -0,0 +1,592 @@
...
+//  Copyright © 2024 OsmAnd. All rights reserved.
+//
+
+import UIKit
+import UniformTypeIdentifiers
+
+@objc(OAMapSettingsGpxViewController)
+@objcMembers
+final class MapSettingsGpxViewController: OABaseNavbarSubviewViewController {
+    private let previouslyVisibleTracksKey: String = "PreviouslyVisibleGpxFilePaths"
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:15]: String - remove

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 3]
```diff
@@ -0,0 +1,592 @@
...
+        self.tableView.setEditing(true, animated: false)
+        self.tableView.allowsMultipleSelectionDuringEditing = true
+        self.navigationItem.hidesSearchBarWhenScrolling = true
+        self.definesPresentationContext = true
+        updateSelectedRows()
+    }
+    
+    override func viewWillDisappear(_ animated: Bool) {
+        super.viewWillDisappear(animated)
+        if let searchController = searchController, searchController.isActive {
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:52]: if let searchController, searchController.isActive {

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 4]
```diff
@@ -0,0 +1,592 @@
...
+            searchController.isActive = false
+            self.navigationItem.searchController = nil
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:54]: self - remove

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 5]
```diff
@@ -0,0 +1,592 @@
...
+    override func getTitle() -> String? {
+        localizedString("shared_string_gpx_tracks")
+    }
+    
+    override func getLeftNavbarButtonTitle() -> String? {
+        localizedString("shared_string_cancel")
+    }
+    
+    override func getRightNavbarButtons() -> [UIBarButtonItem]? {
+        guard let searchButton = createRightNavbarButton(nil, iconName: "ic_navbar_search", action: #selector(onSearchButtonClicked), menu: nil),
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:67]: mb use from assets? UIImage.icNavbarSearch

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 6]
```diff
@@ -0,0 +1,592 @@
...
+        searchButton.accessibilityLabel = localizedString("shared_string_search")
+        importButton.accessibilityLabel = localizedString("import_tracks")
+        return [importButton, searchButton]
+    }
+    
+    override func createSubview() -> UIView? {
+        if !isSearchActive {
+            segmentedControl = UISegmentedControl(items: [localizedString("shared_string_visible"), localizedString("shared_string_all")])
+            segmentedControl?.selectedSegmentIndex = 0
+            isShowingVisibleTracks = true
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:79]: move before  segmentedControl = UISegmentedControl(

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 7]
```diff
@@ -0,0 +1,592 @@
...
+    override func getBottomAxisMode() -> NSLayoutConstraint.Axis {
+        .horizontal
+    }
+    
+    override func getTopButtonTitle() -> String? {
+        if !isTracksAvailable || (isShowingVisibleTracks && !isVisibleTracksAvailable && recentlyVisibleGpxList.isEmpty) {
+            return localizedString("shared_string_select_all")
+        }
+        
+        return areAllTracksSelected() ? localizedString("shared_string_deselect_all") : localizedString("shared_string_select_all")
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:100]: localizedString(reAllTracksSelected() ? "shared_string_deselect_all" :  shared_string_select_all)

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 8]
```diff
@@ -0,0 +1,592 @@
...
+    }
+    
+    override func getBottomButtonTitle() -> String? {
+        localizedString("shared_string_done")
+    }
+    
+    override func getTopButtonColorScheme() -> EOABaseButtonColorScheme {
+        return (!isTracksAvailable || (isShowingVisibleTracks && !isVisibleTracksAvailable && recentlyVisibleGpxList.isEmpty)) ? .inactive : .graySimple
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:108]: move((!isTracksAvailable || (isShowingVisibleTracks && !isVisibleTracksAvailable && recentlyVisibleGpxList.isEmpty)) to additions func (duplicate)

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 9]
```diff
@@ -0,0 +1,592 @@
...
+                    let time = OAOsmAndFormatter.getFormattedTimeInterval(TimeInterval(gpx.timeSpan), shortFormat: true) ?? "N/A"
+                    let waypointCount = "\(gpx.wptPoints)"
+                    let gpxRow = recentlyVisibleSection.createNewRow()
+                    gpxRow.cellType = OASimpleTableViewCell.getIdentifier()
+                    gpxRow.title = title
+                    gpxRow.setObj(distance, forKey: "distance")
+                    gpxRow.setObj(time, forKey: "time")
+                    gpxRow.setObj(waypointCount, forKey: "waypointCount")
+                    gpxRow.iconName = "ic_custom_trip"
+                    gpxRow.iconTintColor = UIColor.iconColorDisabled
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:161]: UIColor.iconColorDisabled -> .iconColorDisabled

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 10]
```diff
@@ -0,0 +1,592 @@
...
+            }
+        } else {
+            let noTracksSection = tableData.createNewSection()
+            let noTracksRow = noTracksSection.createNewRow()
+            noTracksRow.cellType = OALargeImageTitleDescrTableViewCell.getIdentifier()
+            noTracksRow.key = "noTracks"
+            noTracksRow.title = localizedString("no_track_files")
+            noTracksRow.descr = localizedString("import_create_track_files")
+            noTracksRow.iconName = "ic_custom_folder_open"
+            noTracksRow.iconTintColor = UIColor.iconColorDefault
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:172]: .iconColorDefault

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 11]
```diff
@@ -0,0 +1,592 @@
...
+    }
+    
+    override func getRow(_ indexPath: IndexPath?) -> UITableViewCell? {
+        guard let indexPath else { return nil }
+        let item = tableData.item(for: indexPath)
+        var outCell: UITableViewCell?
+        if item.cellType == OASimpleTableViewCell.getIdentifier() {
+            var cell = tableView.dequeueReusableCell(withIdentifier: OASimpleTableViewCell.getIdentifier()) as? OASimpleTableViewCell
+            if cell == nil {
+                let nib = Bundle.main.loadNibNamed(OASimpleTableViewCell.getIdentifier(), owner: self, options: nil)
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:184]: bad practice.
 Add in viewDidLoad
   tableView.register(UINib(nibName: OASimpleTableViewCell.reuseIdentifier, bundle: nil), forCellReuseIdentifier: OASimpleTableViewCell.reuseIdentifier). and remove  let nib = Bundle.main.loadNibNamed(

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 12]
```diff
@@ -0,0 +1,592 @@
...
+                cell = nib?.first as? OASimpleTableViewCell
+            }
+            
+            if let cell {
+                let distance = item.obj(forKey: "distance") as? String ?? ""
+                let time = item.obj(forKey: "time") as? String ?? ""
+                let waypointCount = item.obj(forKey: "waypointCount") as? String ?? ""
+                cell.titleLabel.text = item.title
+                cell.descriptionLabel.text = [distance, time, waypointCount].filter { !$0.isEmpty }.joined(separator: " · ")
+                let iconImage: UIImage
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:194]: let iconName = item.iconName ?? "ic_custom_trip"
let iconImage = UIImage(named: iconName) ?? UIImage()

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 13]
```diff
@@ -0,0 +1,592 @@
...
+                }
+                
+                cell.leftIconView.image = iconImage.withRenderingMode(.alwaysTemplate)
+                cell.leftIconView.tintColor = item.iconTintColor
+            }
+            return cell
+        } else if item.cellType == OALargeImageTitleDescrTableViewCell.getIdentifier() {
+            var cell = tableView.dequeueReusableCell(withIdentifier: OALargeImageTitleDescrTableViewCell.getIdentifier()) as? OALargeImageTitleDescrTableViewCell
+            if cell == nil {
+                let nib = Bundle.main.loadNibNamed(OALargeImageTitleDescrTableViewCell.getIdentifier(), owner: self, options: nil)
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:208]: tableView.register(UINib(nibName

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 14]
```diff
@@ -0,0 +1,592 @@
...
+                cell = nib?.first as? OALargeImageTitleDescrTableViewCell
+                cell?.selectionStyle = .none
+            }
+            
+            if let cell = cell {
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:213]: if let cell {}

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 15]
```diff
@@ -0,0 +1,592 @@
...
+            let update: Bool = outCell?.needsUpdateConstraints() ?? false
+            if update {
+                outCell?.setNeedsUpdateConstraints()
+            }
+        }
+        return outCell
+    }
+    
+    override func onRowSelected(_ indexPath: IndexPath?) {
+        guard let indexPath = indexPath, let gpx = getGpxForSelectedRow(at: indexPath) else { return }
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:239]: guard let indexPath,

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 16]
```diff
@@ -0,0 +1,592 @@
...
+        if !selectedGpxTracks.contains(where: { $0.gpxFilePath == gpx.gpxFilePath }) {
+            selectedGpxTracks.append(gpx)
+        }
+        
+        updateBottomButtons()
+    }
+    
+    override func onRowDeselected(_ indexPath: IndexPath?) {
+        guard let indexPath = indexPath, let gpx = getGpxForSelectedRow(at: indexPath) else { return }
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:248]: guard let indexPath,

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 17]
```diff
@@ -0,0 +1,592 @@
...
+        if let index = selectedGpxTracks.firstIndex(where: { $0.gpxFilePath == gpx.gpxFilePath }) {
+            selectedGpxTracks.remove(at: index)
+        }
+        
+        updateBottomButtons()
+    }
+    
+    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
+        let item = tableData.item(for: indexPath)
+        if item.key == "noVisibleTracks" || item.key == "noTracks" {
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:258]: return !(item.key == "noVisibleTracks" || item.key == "noTracks")

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 18]
```diff
@@ -0,0 +1,592 @@
...
+                self.buttonsBottomOffsetConstraint.constant = 0
+                self.view.layoutIfNeeded()
+            }, completion: nil)
+        }
+    }
+    
+    @objc private func segmentChanged(_ control: UISegmentedControl) {
+        isShowingVisibleTracks = control.selectedSegmentIndex == 0
+        generateData()
+        self.tableView.reloadData()
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:314]: self.

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 19]
```diff
@@ -0,0 +1,592 @@
...
+    }
+    
+    @objc private func onImportButtonClicked() {
+        let contentTypes: [UTType] = [UTType(filenameExtension: "gpx") ?? .item,
+                                      UTType(filenameExtension: "kmz") ?? .item,
+                                      UTType(filenameExtension: "kml") ?? .item]
+        let documentPickerVC = UIDocumentPickerViewController(forOpeningContentTypes: contentTypes, asCopy: true)
+        documentPickerVC.allowsMultipleSelection = false
+        documentPickerVC.delegate = self
+        self.present(documentPickerVC, animated: true, completion: nil)
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:352]: self

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 20]
```diff
@@ -0,0 +1,592 @@
...
+    
+    private func loadVisibleTracks() {
+        guard let visibleGpxFilePaths = OAAppSettings.sharedManager()?.mapSettingVisibleGpx.get() else { return }
+        visibleGpxList = allGpxList.filter { visibleGpxFilePaths.contains($0.gpxFilePath) }
+        isVisibleTracksAvailable = !visibleGpxList.isEmpty
+        selectedGpxTracks = visibleGpxList
+    }
+    
+    private func loadRecentlyVisibleTracks() {
+        if allGpxList.isEmpty {
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:471]: guard

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 21]
```diff
@@ -0,0 +1,592 @@
...
+            selectedGpxTracks.removeAll()
+            for gpx in gpxListToShow where visibleGpxFilePaths.contains(gpx.gpxFilePath) {
+                selectedGpxTracks.append(gpx)
+            }
+        }
+        
+        for (index, gpx) in gpxListToShow.enumerated() {
+            let indexPath = IndexPath(row: index, section: 0)
+            if selectedGpxTracks.contains(where: { $0.gpxFilePath == gpx.gpxFilePath }) {
+                self.tableView.selectRow(at: indexPath, animated: false, scrollPosition: .none)
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:514]: self

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 22]
```diff
@@ -0,0 +1,592 @@
...
+            } else {
+                self.tableView.deselectRow(at: indexPath, animated: false)
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:516]: self.

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 23]
```diff
@@ -0,0 +1,592 @@
...
+            }
+        }
+    }
+    
+    private func hasSelectionChanged() -> Bool {
+        let currentVisibleTrackPaths = visibleGpxList.map { $0.gpxFilePath }
+        let selectedTrackPaths = selectedGpxTracks.map { $0.gpxFilePath }
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:523]: move in additional func selectedGpxTracks.map { $0.gpxFilePath }

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 24]
```diff
@@ -0,0 +1,592 @@
...
+
+extension MapSettingsGpxViewController: UISearchBarDelegate {
+    func searchBarCancelButtonClicked(_ searchBar: UISearchBar) {
+        searchController = nil
+        self.navigationItem.searchController = nil
+        isSearchActive = false
+        isSearchFilteringActive = false
+        filteredGpxList.removeAll()
+        updateNavbar()
+        guard let segmentedControl = segmentedControl else { return }
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:548]: guard let segmentedControl else

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift (hunk 5 above)
Comment (DmitrySvetlichny) [reply Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:67]: The createRightNavbarButton function is designed to accept the icon name as a string.

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift (hunk 15 above)
Comment (DmitrySvetlichny) [reply Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:239]: code must be correct: guard let indexPath = indexPath, let gpx = getGpxForSelectedRow(at: indexPath) else { return }

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift (hunk 16 above)
Comment (DmitrySvetlichny) [reply Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:248]: code must be correct!?

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 25]
```diff
@@ -0,0 +1,570 @@
...
+        
+        return localizedString(areAllTracksSelected() ? "shared_string_deselect_all" : "shared_string_select_all")
+    }
+    
+    override func getBottomButtonTitle() -> String? {
+        localizedString("shared_string_done")
+    }
+    
+    override func getTopButtonColorScheme() -> EOABaseButtonColorScheme {
+        return shouldUseInactiveColorScheme() ? .inactive : .graySimple
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:110]: remove return

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 26]
```diff
@@ -0,0 +1,570 @@
...
+            noTracksRow.title = localizedString("no_track_files")
+            noTracksRow.descr = localizedString("import_create_track_files")
+            noTracksRow.iconName = "ic_custom_folder_open"
+            noTracksRow.iconTintColor = .iconColorDefault
+            noTracksRow.setObj(localizedString("shared_string_import"), forKey: "buttonTitle")
+        }
+    }
+    
+    override func getRow(_ indexPath: IndexPath?) -> UITableViewCell? {
+        guard let indexPath = indexPath else { return nil }
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:180]: guard let indexPath else { return nil }

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift (hunk 15 above)
Comment (tigrim) [reply Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:217]: guard let indexPath, let gpx = getGpxForSelectedRow(at: indexPath) else { return }

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift (hunk 16 above)
Comment (tigrim) [reply Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:226]: up

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 27]
```diff
@@ -483,6 +457,10 @@ final class MapSettingsGpxViewController: OABaseNavbarSubviewViewController {
         recentlyVisibleGpxList = recentlyVisibleTracks
     }
     
+    private func shouldUseInactiveColorScheme() -> Bool {
+        return !isTracksAvailable || (isShowingVisibleTracks && !isVisibleTracksAvailable && recentlyVisibleGpxList.isEmpty)
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:461]: remove return

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 28]
```diff
@@ -0,0 +1,570 @@
...
+        }
+    }
+    
+    @objc private func onSearchButtonClicked() {
+        if searchController == nil {
+            searchController = UISearchController(searchResultsController: nil)
+            searchController?.searchBar.delegate = self
+            searchController?.obscuresBackgroundDuringPresentation = false
+            searchController?.searchBar.placeholder = localizedString("shared_string_search")
+            self.navigationItem.searchController = searchController
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:301]: self

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 29]
```diff
@@ -0,0 +1,570 @@
...
+            let tracksToHide = currentVisibleTrackPaths.compactMap { $0 }.filter { !selectedTrackPaths.contains($0) }
+            for trackPath in tracksToHide {
+                if let track = allGpxList.first(where: { $0.gpxFilePath == trackPath }) {
+                    if !recentlyVisibleGpxList.contains(where: { $0.gpxFilePath == trackPath }) {
+                        recentlyVisibleGpxList.append(track)
+                    }
+                }
+            }
+            
+            recentlyVisibleGpxList.removeAll { track in
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:396]: $0 inline

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 30]
```diff
@@ -0,0 +1,570 @@
...
+        let actionSheet = UIAlertController(title: nil, message: nil, preferredStyle: .actionSheet)
+        let destructiveAction = UIAlertAction(title: localizedString("shared_string_discard_changes"), style: .destructive) { [weak self] _ in
+            self?.dismiss()
+        }
+        
+        let cancelAction = UIAlertAction(title: localizedString("shared_string_cancel"), style: .cancel, handler: nil)
+        actionSheet.addAction(destructiveAction)
+        actionSheet.addAction(cancelAction)
+        if let popoverController = actionSheet.popoverPresentationController {
+            popoverController.barButtonItem = self.navigationItem.leftBarButtonItem
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:417]: self

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 31]
```diff
@@ -0,0 +1,570 @@
...
+            popoverController.permittedArrowDirections = .any
+        }
+        
+        self.present(actionSheet, animated: true, completion: nil)
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:421]: self

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 32]
```diff
@@ -0,0 +1,567 @@
...
+}
+
+extension MapSettingsGpxViewController: UIDocumentPickerDelegate {
+    func documentPicker(_ controller: UIDocumentPickerViewController, didPickDocumentsAt urls: [URL]) {
+        guard let url = urls.first, ["gpx", "kml", "kmz"].contains(url.pathExtension.lowercased()) else { return }
+        let gpxListViewController = OAGPXListViewController()
+        gpxListViewController.prepareProcessUrl(url, showAlerts: true, openGpxView: false) { success in
+            DispatchQueue.main.async {
+                if success {
+                    self.loadGpxTracks()
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:556]: DispatchQueue.main.async { - move down

File: Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift [hunk 33]
```diff
@@ -0,0 +1,567 @@
...
+    }
+    
+    override func viewDidLoad() {
+        super.viewDidLoad()
+        tableView.register(UINib(nibName: OASimpleTableViewCell.getIdentifier(), bundle: nil), forCellReuseIdentifier: OASimpleTableViewCell.getIdentifier())
+        tableView.register(UINib(nibName: OALargeImageTitleDescrTableViewCell.getIdentifier(), bundle: nil), forCellReuseIdentifier: OALargeImageTitleDescrTableViewCell.getIdentifier())
+        tableView.setEditing(true, animated: false)
+        tableView.allowsMultipleSelectionDuringEditing = true
+        navigationItem.hidesSearchBarWhenScrolling = true
+        self.definesPresentationContext = true
```
Comment (tigrim) [inline Sources/Controllers/DashboardOnMap/MapSettings/MapSettingsGpxViewController.swift:48]: sefl

**Review by tigrim — APPROVED** [review]

**Pull request merged**